src/main/java/echoquery/sql/SchemaInferrer.java

```diff
@@ -101,21 +105,55 @@ private void adjustColumnToTable(String table, DatabaseMetaData md)
   }
 
   /**
+   * @param column the column of interest
+   * @param sourceTable the table of interest
+   * @param destinationTable the table being joined
+   * @return which table the column belongs to.
+   */
+  private String inferPrefix(
+      String column, String sourceTable, String destinationTable) {
+    Set<String> tables = columnToTable.get(column);
+    return (tables.contains(sourceTable)) ? sourceTable : destinationTable;
+  }
+
+  private InferredContext inferPrefixes(
+      String aggregation,
+      List<String> comparisons,
+      String table,
+      String destinationTable) {
+    InferredContext ctx = new InferredContext();
+    ctx.setAggregationPrefix(inferPrefix(aggregation, table, destinationTable));
+    for (String comparison : comparisons) {
+      ctx.addComparisonPrefix(inferPrefix(comparison, table, destinationTable));
+    }
+    return ctx;
+  }
+
+  /**
    *
    * @param table the table that we want to aggregate over / select from
    * @param column the column that were filtering on
    * @return a JoinRecipe that can create the table to filter one and select
    *    from
    */
-  public JoinRecipe infer(String table, String column) {
-    // if the inputted table is the center table
-    // find which table has the specific column in columnToTable
-    // find the foreign key to the center table
-    // otherwise, if the column is not in the table give up! if it is you're
-    // good
-    Set<String> criteriaTables =
-        columnToTable.getOrDefault(column, new HashSet<String>());
-    // if the column could be in the table we're querying on
+  public JoinRecipe infer(
+      String table, String aggregation, List<String> comparisons) {
+    // initially we can join with any table
+    Set<String> criteriaTables = new HashSet<>(tables);
+
+    // aggregate all the columns we are interested in
+    List<String> columns = new ArrayList<>(comparisons);
+    columns.add(aggregation);
+
+    // for each column we are interested, we need to filter down only tables
+    // that contain it (ignoring columns that already exist in our base table)
+    for (String column : columns) {
+      Set<String> newCriteria =
+          columnToTable.getOrDefault(column, new HashSet<String>());
+      if (newCriteria.contains(table)) continue;
```

Comment:
I dont think this is the problem- criteriaTables starts having all the columns. if you skip the for loop then criteriaTables will still contain our base table